File: source/javascript/knapsack.js
// 0/1 Knapsack Problem
function knapsack(weights, values, capacity) {
    const n = weights.length;
    const dp = Array.from({ length: n + 1 }, () => new Array(capacity + 1).fill(0));

    for (let i = 1; i <= n; i++) {
        for (let w = 0; w <= capacity; w++) {
            dp[i][w] = dp[i - 1][w];
            if (weights[i - 1] <= w) {
                const withItem = dp[i - 1][w - weights[i - 1]] + values[i - 1];
                dp[i][w] = Math.max(dp[i][w], withItem);
            }
        }
    }
    return dp[n][capacity];
}

const weights = [2, 3, 4, 5];
const values = [3, 4, 5, 6];
const capacity = 8;
console.log(`max value: ${knapsack(weights, values, capacity)}`);

File: source/c/knapsack.c
// gcc -o knapsack knapsack.c
#include <stdio.h>

int main(void) {
    int weights[] = {2, 3, 4, 5};
    int values[] = {3, 4, 5, 6};
    int capacity = 8;
    int n = 4;

    int dp[5][9];
    for (int i = 0; i <= n; i++)
        for (int w = 0; w <= capacity; w++)
            dp[i][w] = 0;

    for (int i = 1; i <= n; i++) {
        for (int w = 0; w <= capacity; w++) {
            dp[i][w] = dp[i-1][w];
            if (weights[i-1] <= w) {
                int with_item = dp[i-1][w - weights[i-1]] + values[i-1];
                if (with_item > dp[i][w]) dp[i][w] = with_item;
            }
        }
    }

    printf("max value: %d\n", dp[n][capacity]);
    return 0;
}
